FirebaseCombineSwift/Tests/Integration/Storage/StorageIntegration.swift

```diff
@@ -0,0 +1,616 @@
+      }
+      .store(in: &cancellables)
+
+    waitForExpectations()
+  }
+
+  func testSimplePutSpecialCharacter() throws {
+    var cancellables = Set<AnyCancellable>()
+    let expectation = self.expectation(description: #function)
+    let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+    let path = "ios/public/-._~!$'()*,=:@&+;"
+    storage.reference(withPath: path)
+      .putData(data)
+      .assertNoFailure()
+      .sink { metadata in
+        XCTAssertEqual(metadata.contentType, "application/octet-stream")
+        expectation.fulfill()
+      }
+      .store(in: &cancellables)
+    waitForExpectations()
+  }
+
+  func testSimplePutDataInBackgroundQueue() throws {
+    var cancellables = Set<AnyCancellable>()
+    let expectation = self.expectation(description: #function)
+    let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+    storage.reference(withPath: "ios/public/testBytesUpload")
+      .putData(data)
+      .subscribe(on: DispatchQueue.global(qos: .background))
+      .assertNoFailure()
+      .sink { _ in
+        expectation.fulfill()
+      }
+      .store(in: &cancellables)
+
+    waitForExpectations()
+  }
+
+  func testSimplePutEmptyData() {
+    var cancellables = Set<AnyCancellable>()
+    let expectation = self.expectation(description: #function)
+    storage
+      .reference(withPath: "ios/public/testSimplePutEmptyData")
+      .putData(Data())
+      .assertNoFailure()
+      .sink { _ in
+        expectation.fulfill()
+      }
+      .store(in: &cancellables)
+    waitForExpectations()
+  }
+
+  func testSimplePutDataUnauthorized() throws {
+    var cancellables = Set<AnyCancellable>()
+    let expectation = self.expectation(description: #function)
+    let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+
+    storage.reference(withPath: "ios/private/secretfile.txt")
+      .putData(data)
+      .filter {
+        print($0)
+        XCTFail("Unexpected success return from putFile)")
+        return false
+      }
+      .mapError { error -> Error in
+        XCTAssertEqual((error as NSError).code, StorageErrorCode.unauthorized.rawValue)
+        expectation.fulfill()
+        return error
+      }
+      .sink { _ in
+      } receiveValue: { _ in
+      }
+      .store(in: &cancellables)
+
+    waitForExpectations()
+  }
+
+  func testAttemptToUploadDirectoryShouldFail() throws {
+    let expectation = self.expectation(description: #function)
+    var cancellables = Set<AnyCancellable>()
+    // This `.numbers` file is actually a directory.
+    let fileName = "HomeImprovement.numbers"
+    let bundle = Bundle(for: StorageIntegration.self)
+    let fileURL = try XCTUnwrap(bundle.url(forResource: fileName, withExtension: ""),
+                                "Failed to get filePath")
+    let ref = storage.reference(withPath: "ios/public/" + fileName)
+
+    ref.putFile(from: fileURL)
```

Comment:
Is there a better way to test for an expected error than this?

Reply:
I've send a PR with a suggestion: #7982

Reply:
Merged and updated other similar error flows. Thanks!